### ISSUE TITLE
Apps have issues connecting using AdGuard VPN

### ISSUE THREAD

**Issue by planet0** [AdguardTeam/AdguardForAndroid#1768]
I'm noticing that with AdGuard enabled, I occasionally keep missing out on any push notifications or apps aren't able to connect to the internet until I bring it up on the screen.

This behaviour stems from reporting that Facebook Messenger isn't sending me new message notifications, and keeps 'waiting for network'. It has now been noticeable with other apps and this morning it now includes Gmail which I'm not happy about, and now believe it's not relating to any filters but with the AdGuard app itself and it's VPN.

- Facebook Messenger not notifying me of new messages, or delays the notification from 10 minutes up to an hour later, saying it is 'Waiting for Network' once I open the app
- WhatsApp tries to run a backup (I have it set to backup on any connection daily), and is always 'Waiting for Network' until I open the app
- Gmail isn't able to retrieve new emails and notify me via notification, and I needed to open the app and manually refresh
- Google Play not notifying me of new updates, need to open the Play store and manual check for updates.

Once I open the apps or force refresh, most of the time it is okay for some time. If I leave my device idle again for a long period of time, the issues noticeably come back, but it is a 'hit and miss' occurrence, as sometimes the issue doesn't arise, and other times the issue comes back (like this morning with Gmail).

If I uninstall AdGuard, then all apps can run in background and give me notifications without a problem.

### Steps to reproduce

1. Leave the phone idle for 6-7 hours (overnight)
2. Unlock the phone to see no notifications
3. Open Messenger, WhatsApp, Gmail and Google Play manually to suddenly receive new notifications and resolve 'Connecting to Network' messages

### Expected behavior
My device can receive push notifications and allow background apps to connect to the Internet.

### Actual behavior
AdGuard prevents background apps from connecting to the Internet until manual user interaction of opening the affected apps.

### Screenshots
<details><summary>Messenger 'Waiting for Network' when Opened:</summary>

![messenger](https://user-images.githubusercontent.com/36845626/37426294-17ac680c-281a-11e8-90bb-75befbf07d70.png)

</details>
<details><summary>Messenger 'Connected' after a few seconds, messages update:</summary>

![messenger-connected](https://user-images.githubusercontent.com/36845626/37426293-176da568-281a-11e8-8356-0168a9db1193.png)

</details>
<details><summary>WhatsApp backup working when User opens phone, otherwise stuck on 'Waiting for Network...':</summary>

![whatsapp-backup](https://user-images.githubusercontent.com/36845626/37426292-17377150-281a-11e8-8fb5-395b61c09198.png)

</details>

### Your environment
* Adguard version: 2.11.31β
* Adguard filtering mode: Local VPN, High-Quality, HTTPS Filtering, DNS Blocking
* Device model: Nokia 6
* Operating system and version: Android 8.0.0
* Rooted or not? No
* Any other apps that can start VPN/Antiviruses/Cleaners etc. installed? No

**Comment by ameshkov** [comment]
> DNS Blocking

Do you use a custom DNS server?

**Comment by ameshkov** [comment]
One more question: check the `pref.vpn.disable.reconfiguration` flag value in the low-level settings

**Comment by planet0** [comment]
This morning, I missed an email from two hours ago. Gmail couldn't notify me until I received another email, then it was able to notify me for both of them. Wasn't too happy about that since it was quite urgent.

> Do you use a custom DNS server?

I selected AdGuard DNS in the DNS server list, and enabled DNS requests blocking.

> One more question: check the pref.vpn.disable.reconfiguration flag value in the low-level settings

That flag is disabled by default. `pref.vpn.disable.pause` is enabled by default, however.

**Comment by zebrum** [comment]
I think I had similar issue some time ago on my old samsung note 2. Every morning after enabling wi-fi
> Gmail couldn't notify me until I received another email,

 But I can't say what was caused it.

**Comment by ameshkov** [comment]
Just in case, does it help to disable battery optimization for AdGuard?

**Comment by planet0** [comment]
> Just in case, does it help to disable battery optimization for AdGuard?
I have considered this. But every time I manually disable battery optimization, it keeps re-enabling itself again at a later point. More noticeably when restarting the device.

**Comment by planet0** [comment]
>I think I had similar issue some time ago on my old samsung note 2. Every morning after enabling wi-fi

Last night I enabled Airplane mode overnight and turned it off this morning, so I got all apps to notify me perfectly. I left my phone alone, and the email arrived in my Inbox around 9:30am, but my phone didn't tell me. I also had a Facebook Message waiting for me that was sent 9:10am. My Facebook Messages keep delaying by 10 minutes, so the app refreshed itself and notified me about the message 9:20am, and my Gmail notification arrived when I got another email at 11:20am, which magically also showed my other email from 9:30am. 

If I manually unlock the phone, some things start to work again, similarly to this mornings Airplane mode enable (manually turning the screen on, unlocking the phone, performing a network change).

I'm hoping all of these details will help paint a picture of what's going on with AdGuard and my phone.

**Comment by ameshkov** [comment]
We need to re-test it on the new beta version which comes this week.

Maybe fixing the #1587 issue will help with this one as well?

**Comment by planet0** [comment]
This might be related to my AdGuard troubles.

This morning on 2.11.67η, Facebook completely stopped connecting via AdGuard. No notifications in background. I disabled AdGuard, and everything instantly worked perfectly. I reenabled AdGuard and it continues to work again. Odd that a manual disable and re-enable continues to 'fix' it temporarily.

Tried to report this via normal means on https://reports.adguard.com/en/new_issue.html,
https://github.com/AdguardTeam/AdguardFilters/issues/12970
https://github.com/AdguardTeam/AdguardFilters/issues/13026

But they were dismissed, presumably as it may be an app issue rather than a filter issue.

<details><summary>Screenshot of Facebook trying to connect:</summary>

![facebook-waiting](https://user-images.githubusercontent.com/36845626/37931681-328912b6-3192-11e8-8698-2583bb884f79.png)

</details>

<details><summary>Screenshot of Facebook telling me it can't connect:</summary>

![facebook-stuck](https://user-images.githubusercontent.com/36845626/37931691-37e9967c-3192-11e8-9f32-0345986ce89b.png)

</details>

**Comment by ameshkov** [comment]
> But they were dismissed, presumably as it may be an app issue rather than a filter issue.

Yeah, this is not a filters issue for sure.

**Comment by ameshkov** [comment]
So have you tried it with the new beta version?

**Comment by planet0** [comment]
I uninstalled AdGuard for a few days to see what would happen, and I had no issues.

>So have you tried it with the new beta version?

I installed 2.11.76β three hours ago. Since then, the Facebook App and Facebook Messenger App have both already stopped giving me notifications, so I've missed new messages and Facebook notifications until I manually accessed the apps.

**Comment by planet0** [comment]
@ameshkov, is there any benefit using both DNS requests blocking and using AdGuard DNS as the custom DNS server? Haven't been able to find any info about this combo.

![screenshot_20180403-111424](https://user-images.githubusercontent.com/36845626/38223898-b5ccb904-3730-11e8-9a7d-fccb24ea378b.png)

**Comment by planet0** [comment]
Today, Facebook thought I was offline at some point whilst the phone was idle. The VPN connection didn't work until I forced refresh a few times in the app. Messenger can't refresh itself in the background.

<details><summary>Screenshots:</summary>

![screenshot_20180404-220506](https://user-images.githubusercontent.com/36845626/38306690-b46be718-3854-11e8-8b0d-662247b6916c.png)

![screenshot_20180404-220518](https://user-images.githubusercontent.com/36845626/38306694-b66ed408-3854-11e8-80be-50a26aa6177d.png)


</details>

**Comment by TPS** [comment]
@planet0 RE: benefit of using both AG local & remote DNS, I believe local is faster, but remote is comprehensive.

I think AGDNS remote, though, should add more features like OpenDNS, like category blocking, smart cache, domain typo resolution, network shortcuts, &c., to make a question like this irrelevant.

**Comment by planet0** [comment]
Thanks @TPS :) I'll keep using both together.

As of now, I've set the following:
- Always-on VPN **enabled** for AdGuard in Android Settings
- Watchdog period set to every **15 seconds**
- perf.enforce.paused.notification **ticked**
- perf.vpn.disable.pause **unticked**
- perf.vpn.disable.reconfigure **unticked**
- perf.vpn.ipv4.force.default **ticked**
- perf.vpn.ipv6.force **ticked**

I really want to get my Android working properly with AdGuard, so I'll see how this setup goes. Any other suggestions or changes to the above, let me know.

**Comment by TPS** [comment]
- Watchdog so high may cause opposite behavior by driving CPU too much.  Even @ my most paranoid/irritated, 60 seconds was more than enough for me.  If you want more, [run Tasker (or similar) on a timer or other conditions, firing intents](https://kb.adguard.com/en/android/solving-problems/tasker).  (I did that w/ great success, even w/ an abandoned precursor to Tasker.)
- The `.force` options are mostly debug.  Leave them off, as AG's heuristics know to work around certain (mostly local, so no ads) network issues when not messed w/.

The rest of the settings are what I use, so they LGTM.

**Comment by planet0** [comment]
Thank you, I wasn't sure about this one, so I'll put it on 60 seconds.
I'll also untick the force options.

**Comment by planet0** [comment]
Bad news, it didn't change anything. Undoing changes and going back.

2.11.76β, Android 8.0.0. Local VPN, High-Quality Filtering, HTTPS Filtering, DNS Blocking and AdGuard DNS. Beta updates. English, Spyware, Social media and Annoyances filters.

<details><summary>Screenshot:</summary>

![screenshot_20180406-173411](https://user-images.githubusercontent.com/36845626/38408581-7bcde672-39c1-11e8-99bc-8290b1d0e480.png)

</details>

**Comment by planet0** [comment]
@ameshkov, @zebrum, @TPS; I'm continuing to have issues with apps all together, no matter if in background or actively using them. I've changed the issue title from background activity, to overall connection issues with the AdGuard VPN.

I'm not really sure why I'm having so much trouble with AdGuard.

Below are screenshots of Instagram unable to connect at all, one error stating **"Failed to send network request"**. Once again, if I disable and re-enable AdGuard, Instagram connects fine again straight away.

<details><summary>News Feed spinning forever:</summary>

![forever-loading](https://user-images.githubusercontent.com/36845626/38421360-3248b8be-39ea-11e8-86b1-916c597167f7.png)


</details><details><summary>Failed to send network request:</summary>

![failed-network-request](https://user-images.githubusercontent.com/36845626/38421376-4175e848-39ea-11e8-8a01-ef811a03f8dc.png)


</details><details><summary>Couldn't refresh feed:</summary>

![refresh-feed](https://user-images.githubusercontent.com/36845626/38421385-4a297fd6-39ea-11e8-9f95-47966b29041b.png)


</details><details><summary>More spinning in any other place in app, this one is the Profile tab:</summary>

![profile-loading](https://user-images.githubusercontent.com/36845626/38421400-55da56b6-39ea-11e8-8d0a-3018d8236cb3.png)


</details>

**Comment by TPS** [comment]
Just some thoughts: What OS & version?  Did you disable AG notification icon?  Is system battery optimization active for AG?

**Comment by TPS** [comment]
Sorry, just saw OS info a few posts up, but I mean: OEM, stock (then what device info?), or some custom ROM?

**Comment by planet0** [comment]
All good @TPS, I'm not too familiar with OS info, but I'll share a screenshot of my 'About phone' section, hope that addresses a lot. It's what's provided by Nokia, I haven't modified anything or loaded anything custom myself.

Also, didn't touch AG notification icon settings, always on. I can manually set battery optimisation off for AG, however it resets back to default ON after every device reboot. I had it on the past 48 hours without rebooting, but the issues were still present, regardless of the battery optimisation setting.

<details><summary>About the Nokia 6</summary>

![screenshot_20180408-065256](https://user-images.githubusercontent.com/36845626/38460304-420a51fc-3afa-11e8-940b-a44a1b7a74a9.png)



</details>

**Comment by TPS** [comment]
I'm not familiar w/ Android 8/Oreo, but I do find it concerning that there're ___NO___ notification icons on your top left, including AG's!  Is that an 8/Oreo thing, or is AG truly not running?

**Comment by planet0** [comment]
>I'm not familiar w/ Android 8/Oreo, but I do find it concerning that there're NO notification icons on your top left, including AG's! Is that an 8/Oreo thing, or is AG truly not running?

That's just a common visual setting, and the default AdGuard setting, 'Notification without icon'. It's the same behaviour with Android 5 and 7 that I've previously experienced, as it still permanently appears in the notification drawer.

![screenshot_20180408-125311](https://user-images.githubusercontent.com/36845626/38462540-36e9d18a-3b2c-11e8-8fb9-8a332059a2ac.png)

**Comment by TPS** [comment]
That screenshot shows that traffic both flowing through & being blocked by AG.  Is your filtering log a solid mass of red, or @ least the entries that have to do w/ your problematic apps?

**Comment by planet0** [comment]
Looking at the Filtering Log, I see very little red entries, only things related to Google Analytics in DNS and some google.com.au in Google Chrome. All other apps (inc. Facebook, Messenger, Instagram) are clear with no red entries.

**Comment by TPS** [comment]
Just to test, have you tried excluding any problematic apps?  Does that help?  Do you have anything else running that has any relation to your data (wifi/mobile) connections?

**Comment by planet0** [comment]
Good idea @TPS, I've only tested either with or without AdGuard installed, but not just excluding the apps in AdGuard.

I've disabled through AdGuard:
- Facebook
- Messenger
- Gmail
- WhatsApp

I'll see how that goes for 24-48 hours and report back.

**Comment by TPS** [comment]
If you have DNS-blocking active, you should get some benefit, anyway.  Other than GMail (special circumstance), are you seeing ads in those apps?

**Comment by planet0** [comment]
New issue popped up, Instagram Stories broke. Refresh doesn't work, can't use any touch controls, have to quit the app. Similar behaviour to when Facebook or Messenger loses its connection, everything just disconnects so harshly and suddenly, like the apps don't even spend any time trying to connect (meaning the connection between the app and the Local VPN fails instantly).

<details><summary>Instagram Story Screenshot</summary>

![screenshot_20180409-092812](https://user-images.githubusercontent.com/36845626/38479175-b54c6072-3c00-11e8-8e61-6b4cd3f8fced.png)
</details>

**Comment by ameshkov** [comment]
@planet0 we'll need to get your "record everything"-level log to figure what's wrong with it.

Along with the log, we'll need to know exact timing - when did the problem occur.

**Comment by planet0** [comment]
This might sound odd, but since enabling "record everything", I haven't had any issues yet.

Is there any plausible reason? I still have the setting set and so far everything is okay. Will see how it goes for the next couple of days.

**Comment by planet0** [comment]
> You still have all those apps excluded, right? 
I don't, as I've reset App Management rules to default.
So this is odd, surely a log setting would be why it's been working fine so far?
Either way, still waiting for an issue. Even Messenger notifications are working instantly.

**Comment by planet0** [comment]
Sent two instances, one with Facebook and one with Messenger, to `devteam at adguard.com`

**Comment by ameshkov** [comment]
Issue 1 (22:19): (Facebook, loading stuck)
I see a sudden disconnection from the Facebook's server side.

Issue 2: 14:05-14:08 (Messenger, no notification for unread message)
It seems that Google Play services app was too busy syncing some Google Drive data right at that moment.

Issue 3: 17:17 (Facebook, something went wrong error message)
I can see that at At 17:16:46.661 Facebook servers closed both active connections

**Comment by ameshkov** [comment]
Could you please try using the system default DNS server for some time?

**Comment by planet0** [comment]
I got fed up with my Nokia device and returned it. Now using a Samsung Galaxy J7 Pro. Will see how this goes for the next couple of days with all the same setup (except DNS will not use AdGuard DNS and instead use 1.1.1.1 and 1.0.0.1).

**Comment by planet0** [comment]
Experiencing the same Facebook Messenger issues with the Samsung device, regardless of AdGuard being installed or not. Also WhatsApp waits for Wi-Fi anyway when connected to Wi-Fi. Issue isn't related to AdGuard for Android, but with the apps themselves.

Sorry for all this trouble.

**Comment by TPS** [comment]
_Out-of-scope & off-topic: @planet0 Well, you changed your device & whether AG is even installed:  Have you debugged your data (wifi/mobile/whatever) connection(s)?  It seems unlikely it's the apps that're causing the issues (unless you're installing from someplace **really** non-standard), since plenty of other folks use them just fine, w/ & w/o AG._

**Comment by planet0** [comment]
> Out-of-scope & off-topic: @planet0 Well, you changed your device & whether AG is even installed: Have you debugged your data (wifi/mobile/whatever) connection(s)?

Hi @TPS, I think I figured out that it was related to AdGuard DNS on my router, as I noticed that out of home and without AdGuard it was functioning fine... and now I'm not using AdGuard DNS on my router (back to ISP default DNS) without any AdGuard products and everything is functional.

So might be to do with AdGuard DNS, or with a different device as well, but either way it's been working as expected the last few days with this Samsung.

**Comment by TPS** [comment]
_@planet0 So, I assume you've 1 of those pretty-common home routers (or ISPs) that don't like downstream devices using a non-default DNS.  If you're still interested in AG, did you try precisely **just** default DNS (w/ or w/o DNS-blocking_ on-device _— the implementation should allow either)?_

**Comment by planet0** [comment]
@TPS I was using an outright TP-Link device that wasn't locked or anything, which I could set any DNS for IPv4 and IPv6. I've also used an ISP specific device as well for the home network.

I needed to go back to using AdGuard, the internet is not a fun place without it. Tried to look at an article on my phone and an ad overtook the browser by redirecting it to a phishing page. 😞 

I'll set all my devices up again to using AdGuard stable on Windows, macOS and Android with AdGuard DNS on the outright TP-Link router and see how it goes.

**Comment by planet0** [comment]
Oh, sorry about that! 😅In the short time I tried with the Nokia device it still was having issues regardless of DNS server settings. 

I no longer have this issue (except for WhatsApp which I now singled out with Issue #1834) as of this Samsung phone with the same original setup, so it was an issue with my previous Nokia device the entire time. It explains why another smartphone (also Samsung) in the house wasn't experiencing issues with AdGuard and my Nokia was.
